test/snapp_auction.js

```diff
@@ -0,0 +1,97 @@
+
+      await truffleAssert.reverts(
+        instance.placeSellOrder(1, 2, 1, 1, { from: user_2 }),
+        "Must have registered account"
+      )
+    })
+
+    it("Reject: buyToken = sellToken", async () => {
+      const instance = await SnappAuction.new()
+      await setupEnvironment(MintableERC20, instance, token_owner, [user_1], 2)
+
+      await truffleAssert.reverts(
+        instance.placeSellOrder(1, 1, 1, 1, { from: user_1 }),
+        "Buy and Sell tokens must differ!"
+      )
+    })
+
+    it("Reject: unregistered buyToken", async () => {
+      const instance = await SnappAuction.new()
+      await setupEnvironment(MintableERC20, instance, token_owner, [user_1], 2)
+
+      await truffleAssert.reverts(
+        instance.placeSellOrder(3, 1, 1, 1, { from: user_1 }),
+        "Buy token is not registered"
+      )
+    })
+
+    it("Reject: unregistered sellToken", async () => {
+      const instance = await SnappAuction.new()
+      await setupEnvironment(MintableERC20, instance, token_owner, [user_1], 2)
+      await truffleAssert.reverts(
+        instance.placeSellOrder(1, 3, 1, 1, { from: user_1 }),
+        "Sell token is not registered"
+      )
+    })
+
+    it("Reject: Buy Amount >= 2^100", async () => {
+      const instance = await SnappAuction.new()
+      await setupEnvironment(MintableERC20, instance, token_owner, [user_1], 2)
+
+      await truffleAssert.reverts(
+        instance.placeSellOrder(1, 2, "0x10000000000000000000000000", 1, { from: user_1 }),
+        "Buy amount too large!"
+      )
+    })
+
+    it("Reject: Sell Amount >= 2^100", async () => {
+      const instance = await SnappAuction.new()
+      await setupEnvironment(MintableERC20, instance, token_owner, [user_1], 2)
+
+      await truffleAssert.reverts(
+        instance.placeSellOrder(1, 2, 1, "0x10000000000000000000000000", { from: user_1 }),
+        "Sell amount too large!"
+      )
+    })
+
+    it("Generic sell order", async () => {
+      const instance = await SnappAuction.new()
+      await setupEnvironment(MintableERC20, instance, token_owner, [user_1], 2)
+      await instance.placeSellOrder(1, 2, 1, 1, { from: user_1 })
+
+      const auctionIndex = (await instance.auctionIndex.call()).toNumber()
+      const currentAuction = await instance.auctions(auctionIndex)
+
+      assert.equal(currentAuction.size, 1)
+      assert.notEqual(currentAuction.shaHash, 0)
+    })
+
+    it("Generic sell orders over two batches", async () => {
+      const instance = await SnappAuction.new()
+      await setupEnvironment(MintableERC20, instance, token_owner, [user_1], 2)
+      await instance.placeSellOrder(1, 2, 1, 1, { from: user_1 })
+
+      await waitForNBlocks(21, owner)
+      await instance.placeSellOrder(1, 2, 1, 1, { from: user_1 })
+
+      const auctionIndex = (await instance.auctionIndex.call()).toNumber()
+      assert.equal(auctionIndex, 1)
+    })
+  })
```

Comment:
I believe in order to get 100% branch coverage, we'd have to test the case where we fill a batch just by pushing a lot of orders in the same block.

Reply:
We already have 100% Branch Coverage. I made sure of this before submitting this PR.

Reply:
Sorry, I believe I meant 100% condition coverage. We are not checking the case where we creating a new batch because the old one is full.

Reply:
It will be difficult to get 1000 orders in one batch before 20 blocks are mined. Any suggestions on how would be very welcome.

Reply:
That's because ganache by default mines a block on every transaction? You could create a smart contract just for the test that opens an account and adds 1000 orders in a single transaction. But not sure if that's worth the hassle.

On a side note, we are using amount of blocks passed rather than an actual measure of time. It's probably somewhat more reliable to use time directly (and create a new batch every 3 minutes). Block times tend to vary significantly and might at some point drastically increase due to the difficulty bomb.
Maybe something to do as part of the final cleanup in this milestone.

Reply:
depending ontime rather than blocks also makes testing most of the time easier... sounds good to me.

Reply:
Created #59